File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/actors/PlayerActorRelocationAdapter.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors;

import com.fasterxml.jackson.databind.ObjectMapper;

import systems.zlink.framework.actors.ZLinkActorRelocationAdapter;
import systems.zlink.framework.actors.ZLinkRelocationCancellation;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class PlayerActorRelocationAdapter
        implements ZLinkActorRelocationAdapter<PlayerActor> {
    private static final ObjectMapper JSON = new ObjectMapper();

    @Override
    public CompletionStage<byte[]> capture(
            PlayerActor actor, ZLinkRelocationCancellation cancellation) {
        try {
            return CompletableFuture.completedFuture(
                    JSON.writeValueAsBytes(
                            new TransferState(
                                    actor.displayName(),
                                    actor.roomId(),
                                    actor.destroyAfterEntrySpotJoin(),
                                    actor.disconnected())));
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    @Override
    public CompletionStage<Void> restore(
            PlayerActor actor, byte[] state, ZLinkRelocationCancellation cancellation) {
        try {
            TransferState transferred = JSON.readValue(state, TransferState.class);
            actor.setDisplayName(transferred.displayName());
            if (transferred.roomId() != null && !transferred.roomId().isBlank()) {
                actor.joinRoom(transferred.roomId());
            }
            if (transferred.destroyAfterEntrySpotJoin()) {
                actor.markForDestroyAfterRoomLeave();
            }
            if (transferred.disconnected()) {
                actor.markDisconnected();
            }
            return CompletableFuture.completedFuture(null);
        } catch (IOException error) {
            return CompletableFuture.failedFuture(error);
        }
    }

    public record TransferState(
            String displayName,
            String roomId,
            boolean destroyAfterEntrySpotJoin,
            boolean disconnected) {}
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/actors/PlayerActorRelocationAdapter.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkActorRelocationAdapter
import systems.zlink.framework.actors.ZLinkRelocationCancellation

class PlayerActorRelocationAdapter : ZLinkActorRelocationAdapter<PlayerActor> {
    override fun capture(
        actor: PlayerActor,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<ByteArray> =
        CompletableFuture.completedFuture(
            json.writeValueAsBytes(
                TransferState(
                    actor.displayName,
                    actor.roomId,
                    actor.destroyAfterEntrySpotJoin,
                    actor.disconnected,
                )
            )
        )

    override fun restore(
        actor: PlayerActor,
        state: ByteArray,
        cancellation: ZLinkRelocationCancellation,
    ): CompletionStage<Void> {
        val transferred = json.readValue(state, TransferState::class.java)
        actor.setDisplayName(transferred.displayName)
        if (transferred.roomId.isNotBlank()) actor.joinRoom(transferred.roomId)
        if (transferred.destroyAfterEntrySpotJoin) actor.markForDestroyAfterRoomLeave()
        if (transferred.disconnected) actor.markDisconnected()
        return CompletableFuture.completedFuture(null)
    }

    companion object {
        private val json = jacksonObjectMapper()
    }

    data class TransferState(
        val displayName: String,
        val roomId: String,
        val destroyAfterEntrySpotJoin: Boolean,
        val disconnected: Boolean,
    )
}
